Set and unset firebase token with account name

n13 opened the pull request:
### 🗃 Github Issue Or Explanation for this PR. (What is it supposed to do and Why is needed)

#1136 


### ✅ Checklist

- [x] I have tested all my changes.

### 🕵️‍♂️ Notes for Code Reviewer

I was trying to come up with a fool proof way to set and unset the firebase token. 

I put it in the setter in settingsStorage, but @gguijarro-c-chwy disagreed and said it should be use cases

So I made use cases, which are now being used instead of direct calls into the settingStorage 

I made _accountName private so nobody can change it outside settingsStorage

This bug can now no longer happen

### 🙈 Screenshots
### 👯‍♀️ Paired with

n13 committed:
set and unset firebase token with account name

n13 committed:
fix lint issues

n13 committed:
remove change of enableRecoveryMode

n13 committed:
firebase user ID can't be null

gguijarro-c-chwy commented on lib/datasource/local/settings_storage.dart:
```diff
@@ -208,6 +215,15 @@ class _SettingsStorage {
     return format.currencyName ?? currencyDefaultCode;
   }
+
+  void _updateFirebaseToken(String oldAccount, String newAccount) {
```
The setting storage should have no knowledge of firebase FirebaseMessageTokenRepository().

gguijarro-c-chwy commented on lib/blocs/authentication/viewmodels/authentication_bloc.dart:
```diff
@@ -27,14 +26,10 @@ class AuthenticationBloc extends Bloc<AuthenticationEvent, AuthenticationState>
-      await FirebaseMessageTokenRepository().setFirebaseMessageToken(settingsStorage.accountName);
     }
     if (event is EnablePasscode) {
       settingsStorage.savePasscode(event.newPasscode);
```
the correct way to implement this is to have a use case that handles these things. 
The settingsStorage should not be doing side effect things like `_updateFirebaseToken` 

The settings storage should know only about settings.

n13 replied:
IDK what to do about this

Use case? And how can we make sure the use case always responds to the account name?

Then we're back in the situation someone might use that use case at the wrong time etc. 

We want to set the firebase token whenever the account name is changed. 

In reality, what we could do, is fire an event when the account name changes. That way settings doesn't need to know about firebase (which makes sense, I agree on this point)

n13 replied:
Ok did the use case. It's good.

n13 committed:
change to event bus

n13 committed:
Merge branch 'bugfix/firebase_token_refactor' of https://github.com/JoinSEEDS/seeds_light_wallet into bugfix/firebase_token_refactor

n13 committed:
changed from event bus to use case

n13 committed:
add await for remove account since it waits for settings to be cleared

we don't want to wait for firebase tokens since we don't want to delay logout just to clear up tokens (network calls)

n13 committed:
cleanup

n13 committed:
cleanup

n13 commented on lib/screens/authentication/recover/recover_account_found/interactor/usecases/fetch_recover_guardian_initial_data.dart:
```diff
@@ -21,7 +22,8 @@ class FetchRecoverGuardianInitialDataUseCase {
       recoveryPrivateKey = settingsStorage.privateKey!;
     } else {
       recoveryPrivateKey = EOSPrivateKey.fromRandom().toString();
-      settingsStorage.enableRecoveryMode(accountName: accountName, privateKey: recoveryPrivateKey);
```
we could make a use case for enable recovery mode - but since it was the exact same as save account plus that flag - I don't think it adds to the code?!

LMK what you think

RaulUrtecho replied:
I like your solution with the use cases, I would also include the flag (privateKeyBackedUp) in them.
They give more clarity than creating functions in the settings storage

gguijarro-c-chwy replied:
i would add a use case. 
calling repos directly should be avoided. Use cases are the preferred way

n13 replied:
Ok done

n13 committed:
use use case for save account

RaulUrtecho commented on lib/domain-shared/shared_use_cases/save_account_use_case.dart:
```diff
@@ -0,0 +1,40 @@
+    final String oldAccountName = settingsStorage.accountName;
+    await settingsStorage.removeAccount();
+    unawaited(updateFirebaseToken(oldAccount: oldAccountName));
+  }
```
There is no something like an Ignore await comment?
`// ignore: unawaited_future`
I comment this because it adds more dependence on Pedantic, if we were thinking to change the lint pacakge in the future. https://github.com/JoinSEEDS/seeds_light_wallet/pull/1121#issuecomment-905893599

n13 replied:
Thanks @RaulUrtecho I like this better too - didn't know I could do that

n13 replied:
Done

gguijarro-c-chwy commented on lib/domain-shared/shared_use_cases/save_account_use_case.dart:
```diff
@@ -0,0 +1,40 @@
+  }
+}
+
+class SaveAccountUseCase extends AccountUseCase {
```
pls move each use case to its own file. Consistency.

n13 replied:
Done

n13 committed:
Merge branch 'master' into bugfix/firebase_token_refactor

n13 committed:
lint complaints

n13 committed:
moved use cases into their own files, added use case for recovery

n13 committed:
fix filename

n13 committed:
using use cases everywhere

n13 committed:
move method

n13 commented on lib/screens/authentication/recover/recover_account_found/interactor/mappers/fetch_recover_guardian_state_mapper.dart:
```diff
@@ -50,7 +50,7 @@ class FetchRecoverRecoveryStateMapper extends StateMapper {
       }
 
       // Save the private key and account
```
This is redundant - the account and key have already been saved when the recovery was initiated.

Recovery process is pretty confusing. 

Anyway, I am not changing this code, it was saving these two items before and now still saving them. 

But - we don't need this and it is a bit of a risk the state got out of sync.

Recovery process should be fixed up on a future PR.

RaulUrtecho replied:
Yeah I also think that when I saw that saving instruction.

n13 committed:
Restore changes in master that were removed in merge

n13 committed:
cleanup

n13 committed:
cleanup

n13 committed:
undoing renames

n13 commented on lib/blocs/authentication/viewmodels/authentication_bloc.dart:
```diff
@@ -23,27 +24,24 @@ class AuthenticationBloc extends Bloc<AuthenticationEvent, AuthenticationState>
-      await FirebaseMessageTokenRepository().setFirebaseMessageToken(event.account);
     }
     if (event is OnRecoverAccount) {
-      settingsStorage.saveAccount(event.account, event.privateKey);
```
This is unrelated to this PR should be fixed up in another PR, but it's not damaging or causing errors, just a little weird

n13 commented on lib/datasource/local/settings_storage.dart:
```diff
@@ -168,27 +168,22 @@ class _SettingsStorage {
-    this.privateKey = privateKey;
-  }
-
-  void saveAccountFromWaitingForRecover(String accountName, String privateKey) {
```
@RaulUrtecho I removed this method again since it's the same as saveAccount, and it's also redundant - see other comment

RaulUrtecho replied:
Agree

gguijarro-c-chwy commented on lib/domain-shared/shared_use_cases/account_use_case.dart:
```diff
@@ -0,0 +1,14 @@
+import 'package:seeds/datasource/remote/firebase/firebase_message_token_repository.dart';
+
+abstract class AccountUseCase {
+  Future<void> updateFirebaseToken({required String oldAccount, String newAccount = ""}) async {
```
this newAccount should be nullable instead of empty, and check for nullability

RaulUrtecho replied:
But what if it recive an empty string? Is possible an empty string arrives here? I think on that was based the latest bug sky found.

gguijarro-c-chwy replied:
i see.Then dont pre set to ""
String newAccount = ""

remove the default

n13 replied:
Huh? The standard is that newAccount = "" when it's empty. That's what has been returned from the settings too, the settings never return null. 

I don't understand this.

In our previous agreement settingStorage either has an account name or it has ""

Changing this for this PR is dangerous and introduces problems. 

Given that it is always set in settingStorage, I find it weird to pretend like we're setting it to null here.

settingStorage returns "" when it's set to null.

n13 replied:
Trying to fix a bug here, not change the entire code base - this topic unrelated to this PR.

gguijarro-c-chwy commented on lib/domain-shared/shared_use_cases/account_use_case.dart:
```diff
@@ -0,0 +1,14 @@
+
+abstract class AccountUseCase {
+  Future<void> updateFirebaseToken({required String oldAccount, String newAccount = ""}) async {
+    if (oldAccount != newAccount) {
```
then this check is 

if(newAccount == null) {
}

n13 replied:
Ok but why is this blocking the PR??

gguijarro-c-chwy commented on lib/domain-shared/shared_use_cases/account_use_case.dart:
```diff
@@ -0,0 +1,14 @@
+import 'package:seeds/datasource/remote/firebase/firebase_message_token_repository.dart';
+
+abstract class AccountUseCase {
```
I dont love this abstraction 

updateFirebaseToken - RemoveToken or Set Token. Neither of the actions is an update. 

I would remove this abstraction and just call removeFirebaseMessageToken or setFirebaseMessageToken as needed in the use case.

n13 replied:
Then we duplicate this code all over the place?

This is an important bugfix we need to merge into the code - we can't block this all the time.

n13 replied:
updateFirebaseToken is responsible for making sure we do the right thing whenever the account name changes. 

I had this as part of set account, which IMO maybe actually better since it ensures this bug this PR is fixing can never happen again.

Use cases are a little more error prone and less secure, but also less entangled, so I can see how they are OK to use as well. 

But in reality when I fix a bug the best outcome is that programmer error can no longer cause the bug. That's what it's aiming for.

gguijarro-c-chwy replied:
I'll review again in 1 hour. Having dinner

RaulUrtecho commented on lib/blocs/authentication/viewmodels/authentication_bloc.dart:
```diff
@@ -23,27 +24,24 @@ class AuthenticationBloc extends Bloc<AuthenticationEvent, AuthenticationState>
     }
     if (event is OnRecoverAccount) {
-      settingsStorage.saveAccount(event.account, event.privateKey);
+      // TODO(RaulUrtecho): - This is redundant - user was already saved when recovery was started.
+      // also a recovery is 1:1 tied to a user account so the account can't change
+      // the key we use as argument here comes from settingsStorage and goes back there.
+      SaveAccountUseCase().run(accountName: event.account, privateKey: event.privateKey);
```
So the account and key both are already saved previous reach this event. if so you right no make sense save again here.
